text/0000-add-builder-init.md

```diff
@@ -0,0 +1,197 @@
+
+```
+import time
+
+print("Starting an infinite loop. Ignores SIGTERM.")
+
+while True:
+    time.sleep(1)
+    print("Still running...")
+```
+
+Given the above application we can:
+
+1. Create an image using `pack build eg1`
+2. Run using podman/docker `docker run –rm -it eg1`
+3. Notice that a `podman stop` or `docker stop -t 10` does not immediately stop the running image and resorts to a `SIGKILL`
+
+The following example creates a single zombie child process. The container will exit after `60` seconds. However, until the container exits the zombie process occupies an entry in the process table. In extreme cases this can exhaust all entries in the process table.
+
+```
+import os
+import time
+
+pid = os.fork()
+
+if pid == 0:
+    # Child process
+    print(f"Child PID {os.getpid()}...")
+    os._exit(0)
+else:
+    # Parent process: sleep without calling wait()
+    print(f"Parent PID {os.getpid()}, child PID {pid}")
+    print("Not waiting for child to finish (zombie will be created)")
+    time.sleep(60)
+```
+
+Given the above example we can:
+
+1. Create an image using `pack build eg2`
+2. Run `eg2` using `podman` or `docker`
+3. Observe that `docker exec {container_id} ps` shows that the child process is a zombie
+
+The two motivating problems of signal handling and sub-process reaping can be addressed either by pushing the respoinsiblity for signal handling and sub-process reaping back onto developers, or by providing platform operators with an option to include an init-style system. This RFC allows platoform-operators to take control of the problem and proposes a mechanism allowing them to opt-in to an init-syle system on their platform.
+
+# What it is
+[what-it-is]: #what-it-is
+
+We propose to allow platform operators define a subreaper for all images built using their builder. The use of a subreaper is optional. The omission of a subreaper from the `builder.toml` defaults to the current behaviour.
+
+We propose that the subreaper, for example `tini`, is launched as PID 1 and runs the process defined in the image.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Assuming an `init` binary, `tini` in this example, is available on the run time, we can simulate how processes can be forked from the `init` binary:
+
+`docker run --rm -it --entrypoint /bin/tini py -s -- /cnb/process/web`
+
+We propose that the enablement of a subreaper is configured at the builder level. This allows platform operators to control whether or not they want to use a subreaper on their platform.
+
+```
+[init]
+enabled = true
+binary = /bin/tini
+```
+
+Builders for which `init` is enabled will result in the following `launch.toml` being produced. Where `init` is provided in `launch.toml` the entrypoint for the exported image will use the given init binary as PID1.
+
+```
+[[processes]]
+type = "web"
+init = /bin/tini
```

Comment:
What's your thinking on putting `init =` with every entry vs an `[init]` section in `launch.toml` and allowing launcher to codify the `direct` behavior?

Reply:
ah ok, I missed the each process could change the behavior part.